src/main/java/com/splunk/kafka/connect/SplunkSinkTask.java

```diff
@@ -250,39 +288,98 @@ private Event createHecEventFrom(final SinkRecord record) {
         if (connectorConfig.raw) {
             RawEvent event = new RawEvent(record.value(), record);
             event.setLineBreaker(connectorConfig.lineBreaker);
+            if(connectorConfig.headerSupport) { event = (RawEvent)addHeadersToHecEvent(event, record); }
             return event;
         }
 
-        // meta data for /event endpoint is per event basis
-        JsonEvent event = new JsonEvent(record.value(), record);
-        if (connectorConfig.useRecordTimestamp && record.timestamp() != null) {
-            // record timestamp is in milliseconds
-            event.setTime(record.timestamp() / 1000.0);
+        JsonEvent event = null;
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        if(connectorConfig.hecEventFormatted) {
+            try {
+                event = objectMapper.readValue(record.value().toString(), JsonEvent.class);
+                log.debug("event is of correct format - {}", event.toString());
+            } catch(Exception e) {
+                log.error("event does not follow correct HEC pre-formatted format", record.toString());
+                event = createHECEventNonFormatted(record);
+            }
+        }
+        else {
+            event = createHECEventNonFormatted(record);
         }
 
-        Map<String, String> metas = connectorConfig.topicMetas.get(record.topic());
-        if (metas != null) {
-            event.setIndex(metas.get(SplunkSinkConnectorConfig.INDEX));
-            event.setSourcetype(metas.get(SplunkSinkConnectorConfig.SOURCETYPE));
-            event.setSource(metas.get(SplunkSinkConnectorConfig.SOURCE));
-            event.addFields(connectorConfig.enrichments);
+        if(connectorConfig.headerSupport) {
+            addHeadersToHecEvent(event, record);
         }
 
         if (connectorConfig.trackData) {
-            // for data loss, latency tracking
             Map<String, String> trackMetas = new HashMap<>();
             trackMetas.put("kafka_offset", String.valueOf(record.kafkaOffset()));
             trackMetas.put("kafka_timestamp", String.valueOf(record.timestamp()));
             trackMetas.put("kafka_topic", record.topic());
             trackMetas.put("kafka_partition", String.valueOf(record.kafkaPartition()));
             event.addFields(trackMetas);
         }
-
         event.validate();
 
         return event;
     }
 
+    private Event addHeadersToHecEvent(Event event, SinkRecord record) {
+        //TODO: Will need exception handling around String serialization
+        Headers headers = record.headers();
+        log.debug(headers.toString());
+        if (!headers.isEmpty()) {
```

Comment:
shortcut this function by doing flip logic and you will have less nested if else
```
if (headers.isEmpty()) {
    return record;
}

...
```

Reply:
Modified to

`    private Event addHeaders(Event event, SinkRecord record) {
        Headers headers = record.headers();

        if(headers.isEmpty() &&  connectorConfig.headerCustom.isEmpty()) {
            return event;
        }

        if (headers.lastWithName(connectorConfig.headerIndex) != null) {
            log.debug("splunk_index header detected, value is: " + headers.lastWithName(connectorConfig.headerIndex).value().toString());
            event.setIndex(headers.lastWithName(connectorConfig.headerIndex).value().toString());
        }
        if (headers.lastWithName(connectorConfig.headerHost) != null) {
            log.debug("splunk_host header detected, value is: " + headers.lastWithName(connectorConfig.headerHost).value().toString());
            event.setHost(headers.lastWithName(connectorConfig.headerHost).value().toString());
        }
        if (headers.lastWithName(connectorConfig.headerSource) != null) {
            log.debug("splunk_source header detected, value is: " + headers.lastWithName(connectorConfig.headerSource).value().toString());
            event.setSource(headers.lastWithName(connectorConfig.headerSource).value().toString());
        }
        if (headers.lastWithName(connectorConfig.headerSourcetype) != null) {
            log.debug("splunk_sourcetype header detected, value is: " + headers.lastWithName(connectorConfig.headerSourcetype).value().toString());
            event.setSourcetype(headers.lastWithName(connectorConfig.headerSourcetype).value().toString());
        }

        if (!connectorConfig.headerCustom.isEmpty()) {
            String[] customHeaders = connectorConfig.headerCustom.split(",");
            Map<String, String> headerMap = new HashMap<>();
            for (String header : customHeaders) {
                if (headers.lastWithName(header) != null) {
                    log.debug(header + " header detected, value is: " + headers.lastWithName(header).value().toString());
                    headerMap.put(header, headers.lastWithName(header).value().toString());
                } else {
                    log.debug(header + " header value not present in event.");
                }
            }
            event.addFields(headerMap);
        }
        return event;
    }`